segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+            // understand the boundaries of events in the absence of an actual notion of event.
+            return segment.append(BufferView.builder().add(intToByteArraySegment(event.getLength())).add(event).build(), null, timeout)
+                    .thenApply(offset -> outstandingBytes.addAndGet(event.getLength() + Integer.BYTES));
+        }
+
+        @Override
+        public void close() {
+            metrics.close();
+        }
+    }
+
+    @Override
+    public void close() {
+        log.info("Closing ContainerEventProcessor service.");
+        closed.set(true);
+    }
+
+    //endregion
+
+    //region AbstractService Implementation
+
+    @Override
+    protected void shutDown() throws Exception {
+        log.info("Shutting down ContainerEventProcessor service.");
+        eventProcessorMap.forEach((k, v) -> v.close());
+        eventProcessorMap.clear();
+        log.info("Shut down of ContainerEventProcessor service complete.");
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        // A ContainerEventProcessor iteration is made of the following stages (for each EventProcessor registered):
+        // 1. Read the LAST_PROCESSED_EVENT_OFFSET attribute to know the starting read offset and read at most getMaxItemsAtOnce events.
+        // 2. The collected results are passed to the handler function in EventProcessor.
+        // 3. If the handler function has been successfully executed, truncate the internal Segment of the EventProcessor
+        // according to the last successfully processed event. If an error occurs, throw and re-try.
+        // 4. Report metrics for each EventProcessor registered and wait for ITERATION_SLEEP until the next iteration.
+        log.info("Starting ContainerEventProcessor service.");
+        Futures.loop(
+                () -> !closed.get(),
+                () -> Futures.delayedFuture(ITERATION_SLEEP, this.executor)
+                             .thenRunAsync(this::processEvents, this.executor)
+                             .thenRunAsync(this::reportMetrics, this.executor),
+                executor);
+    }
+
+    /**
+     * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each EventProcessor:
+     * i) getting the last processed event offset, ii) read at most {@link EventProcessorConfig#getMaxItemsAtOnce()},
+     * iii) execute the handler function for the {@link EventProcessor}, and iv) truncate the internal Segment.
+     */
+    private CompletableFuture<Void> processEvents() {
+        final Timer iterationTime = new Timer();
+        List<CompletableFuture<Object>> futures = new ArrayList<>();
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            futures.add(readEventsForEventProcessor(ep)
+                    .thenApplyAsync(readResult -> applyProcessorHandler(ep, readResult), this.executor)
```

Comment:
Combine all these into a single handler - no need to chain these like this (too many callbacks on the executor).